benchmarks/dataset_serialize_benchmark.py

```diff
@@ -0,0 +1,272 @@
+import itertools
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import time
+import uuid
+
+import conbench.runner
+import pyarrow
+import pyarrow.dataset as ds
+
+from benchmarks import _benchmark
+
+log = logging.getLogger(__name__)
+
+
+# All benchmark scnearios will write below /dev/shm/<SHM_DIR_PREFIX>. That
+# directory tree is removed upon completion (not necessarily in case of error
+# though).
+OUTPUT_DIR_PREFIX = os.path.join("/dev/shm/", "bench-" + str(uuid.uuid4())[:8])
+
+
+@conbench.runner.register_benchmark
+class DatasetSerializeBenchmark(_benchmark.Benchmark):
+    """
+    This benchmark is supposed to measure the time it takes to write data from
+    memory (from a pyarrow Table) to a tmpfs file system, given a specific
+    serialization format (parquet, arrow, ...).
+
+    To make this benchmark agnostic to disk read performance on the input side
+    of things, the data is read fully into memory before starting (and timing)
+    the benchmark function. That is (believed to be) achieved with:
+
+        data = source_dataset.to_table(filter=somefilter, ...)
+
+    After data of interest has been read into memory, the following call is
+    used for both serialization and writing-to-filesystem in one go:
+
+        pyarrow.dataset.write_dataset(format=someformat, ...)
+
+    That operation is timed (and the duration is the major output of this
+    benchmark).
+
+    The data is written to `/dev/shm` (available on all Linux systems). This is
+    a file system backed by RAM (tmpfs). The assumption is that writing to
+    tmpfs is fast (so fast that benchmark duration is significantly affected by
+    the serialization itself), and stable (so that its performance is ~constant
+    across runs on the same machine).
+
+    This benchmark does not resolve how much time goes into the CPU work for
+    serialization vs. the system calls for writing to tmpfs (that would be a
+    different question to answer, an interesting one, that is maybe more of a
+    task for profiling).
+
+    There are two dimensions that are varied:
+
+        - serialization format
+        - amount of the data being written, as set by a filter on the input
+
+    A note about /dev/shm: it's of great value because
+
+    - unprivileges users can write to it
+    - the `base_dir` arg to pyarrow.dataset.write_dataset() requires a path to
+      a directory. That is, one cannot inject a memory-backed Python file
+      object (a strategy that's elsewhere often used to simulate writing to an
+      actual file)
+    - it is not available on MacOS which is why we skip this test
+
+    """
+
+    name = "dataset-serialize"
+
+    arguments = ["source"]
+
+    sources = [
+        "nyctaxi_multi_parquet_s3",
+        "nyctaxi_multi_ipc_s3",
+        # "chi_traffic_2020_Q1",
```

Comment:
With the `chi_...` benchmark I do get:

```
[221214-16:26:00.151] [370389] [root] ERROR: {"timestamp": "2022-12-14T15:26:00.151675+00:00", "tags": {"dataset": "chi_traffic_2020_Q1", "cpu_count": null, "selectivity": "10pc", "format": "csv", "name": "dataset-serialize"}, "info": {"arrow_version": "10.0.1", "arrow_compiler_id": "GNU", "arrow_compiler_version": "10.2.1", "benchmark_language_version": "Python 3.10.8"}, "context": {"arrow_compiler_flags": " -fdiagnostics-color=always -O2 -DNDEBUG -ftree-vectorize", "benchmark_language": "Python"}, "error": "Unsupported Type:struct<sec: double, min: int32, hour: int32, mday: int32, mon: int32, year: int32, wday: int32, yday: int32, isdst: int32, zone: string, gmtoff: int32>"}
Traceback (most recent call last):
  File "/home/jp/dev/voltrondata-labs-benchmarks/benchmarks/_benchmark.py", line 86, in benchmark
    benchmark, output = self.conbench.benchmark(
  File "/home/jp/dev/voltrondata-labs-benchmarks/conbench/conbench/runner.py", line 188, in benchmark
    raise e
  File "/home/jp/dev/voltrondata-labs-benchmarks/conbench/conbench/runner.py", line 160, in benchmark
    data, output = self._get_timing(f, iterations, timing_options)
  File "/home/jp/dev/voltrondata-labs-benchmarks/conbench/conbench/runner.py", line 363, in _get_timing
    output = f()
  File "/home/jp/dev/voltrondata-labs-benchmarks/benchmarks/dataset_serialize_benchmark.py", line 242, in benchfunc
  File "/home/jp/.pyenv/versions/3108-vd-benchmarks/lib/python3.10/site-packages/pyarrow/dataset.py", line 988, in write_dataset
    _filesystemdataset_write(
  File "pyarrow/_dataset.pyx", line 2859, in pyarrow._dataset._filesystemdataset_write
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Unsupported Type:struct<sec: double, min: int32, hour: int32, mday: int32, mon: int32, year: int32, wday: int32, yday: int32, isdst: int32, zone: string, gmtoff: int32>
```

Reply:
@jonkeane should we submit this as an issue somewhere?

Reply:
Looking at that error, there's an issue converting a struct to a text representation for the csv representation. There might be an issue in apache/arrow for that already (though I suspect that faithful writing of structs to csvs is likely a feature that apache/arrow will punt on or possibly already did punt on)